src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,748 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\String;
+
+use Symfony\Component\String\Exception\ExceptionInterface;
+use Symfony\Component\String\Exception\InvalidArgumentException;
+
+/**
+ * Represents a string.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Hugo Hamon <hugohamon@neuf.fr>
+ *
+ * @throws ExceptionInterface
+ */
+abstract class AbstractString implements \JsonSerializable
+{
+    protected $string = '';
+    protected $ignoreCase = false;
+
+    abstract public function __construct(string $string = '');
+
+    /**
+     * @param string|string[] $needle
+     *
+     * @return static
+     */
+    public function after($needle, bool $includeNeedle = false): self
+    {
+        if (null === $i = $this->indexOf($needle)) {
+            $str = clone $this;
+            $str->string = '';
+
+            return $str;
+        }
+
+        if (!$includeNeedle) {
+            $str = clone $this;
+            $str->string = $needle;
+            $i += $str->length();
+        }
+
+        return $this->slice($i);
+    }
+
+    /**
+     * @param string|string[] $needle
+     *
+     * @return static
+     */
+    public function afterLast($needle, bool $includeNeedle = false): self
+    {
+        if (null === $i = $this->indexOfLast($needle)) {
+            $str = clone $this;
+            $str->string = '';
+
+            return $str;
+        }
+
+        if (!$includeNeedle) {
+            $str = clone $this;
+            $str->string = $needle;
+            $i += $str->length();
+        }
+
+        return $this->slice($i);
+    }
+
+    /**
+     * @return static
+     */
+    abstract public function append(string ...$suffix): self;
+
+    /**
+     * @param string|string[] $needle
+     *
+     * @return static
+     */
+    public function before($needle, bool $includeNeedle = false): self
+    {
+        if (null === $i = $this->indexOf($needle)) {
+            $str = clone $this;
+            $str->string = '';
+
+            return $str;
+        }
+
+        if ($includeNeedle) {
+            $str = clone $this;
+            $str->string = $needle;
+            $i += $str->length();
+        }
+
+        return $this->slice(0, $i);
+    }
+
+    /**
+     * @param string|string[] $needle
+     *
+     * @return static
+     */
+    public function beforeLast($needle, bool $includeNeedle = false): self
+    {
+        if (null === $i = $this->indexOfLast($needle)) {
+            $str = clone $this;
+            $str->string = '';
+
+            return $str;
+        }
+
+        if ($includeNeedle) {
+            $str = clone $this;
+            $str->string = $needle;
+            $i += $str->length();
+        }
+
+        return $this->slice(0, $i);
+    }
+
+    /**
+     * @return static[]
+     */
+    abstract public function chunk(int $length = 1): array;
+
+    /**
+     * @param string|string[] $needle
+     */
+    public function contains($needle): bool
```

Comment:
How do you decide on functions like this to use or not a variadic argument?

Reply:
That's not trivial - extensibility maybe. variadics are unexpected to me most of the time.

Reply:
> extensibility maybe.

ok.. i see. you mean it will block you from adding an extra argument if needed...